quickwit/quickwit-doc-mapper/src/tokenizers.rs

```diff
@@ -17,34 +17,279 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
+use std::str::CharIndices;
+
 use once_cell::sync::Lazy;
-use tantivy::tokenizer::{RawTokenizer, RemoveLongFilter, TextAnalyzer, TokenizerManager};
+use tantivy::tokenizer::{
+    BoxTokenStream, LowerCaser, RawTokenizer, RemoveLongFilter, TextAnalyzer, Token, TokenStream,
+    Tokenizer, TokenizerManager,
+};
 
 fn get_quickwit_tokenizer_manager() -> TokenizerManager {
     let raw_tokenizer = TextAnalyzer::from(RawTokenizer).filter(RemoveLongFilter::limit(100));
 
+    let chinese_tokenizer = TextAnalyzer::from(ChineseTokenizer)
+        .filter(RemoveLongFilter::limit(40))
+        .filter(LowerCaser);
+
     let tokenizer_manager = TokenizerManager::default();
+
     tokenizer_manager.register("raw", raw_tokenizer);
+    tokenizer_manager.register("chinese", chinese_tokenizer);
+
     tokenizer_manager
 }
 
+#[derive(Clone)]
+struct ChineseTokenizer;
+
+impl Tokenizer for ChineseTokenizer {
+    fn token_stream<'a>(&self, text: &'a str) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(ChineseTokenStream {
+            text,
+            last_char: None,
+            chars: text.char_indices(),
+            token: Token::default(),
+        })
+    }
+}
+
+struct ChineseTokenStream<'a> {
+    text: &'a str,
+    last_char: Option<(usize, char)>,
+    chars: CharIndices<'a>,
+    token: Token,
+}
+
+fn char_is_cjk(c: char) -> bool {
+    // Block                                   Range       Comment
+    // CJK Unified Ideographs                  4E00-9FFF   Common
+    // CJK Unified Ideographs Extension A      3400-4DBF   Rare
+    // CJK Unified Ideographs Extension B      20000-2A6DF Rare, historic
+    // CJK Unified Ideographs Extension C      2A700–2B73F Rare, historic
+    // CJK Unified Ideographs Extension D      2B740–2B81F Uncommon, some in current use
+    // CJK Unified Ideographs Extension E      2B820–2CEAF Rare, historic
+    matches!(c,
+        '\u{4500}'..='\u{9FFF}' |
+        '\u{3400}'..='\u{4DBF}' |
+        '\u{20000}'..='\u{2A6DF}' |
+        '\u{2A700}'..='\u{2CEAF}' // merge of extension C,D and E.
+    )
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum Grouping {
+    Keep,
+    SplitKeep,
+    SplitIgnore,
+}
+
+fn char_grouping(c: char) -> Grouping {
+    if c.is_alphanumeric() {
+        if char_is_cjk(c) {
+            Grouping::SplitKeep
+        } else {
+            Grouping::Keep
+        }
+    } else {
+        Grouping::SplitIgnore
+    }
+}
+
+impl<'a> TokenStream for ChineseTokenStream<'a> {
+    fn advance(&mut self) -> bool {
+        self.token.text.clear();
+        self.token.position = self.token.position.wrapping_add(1);
+
+        let mut iter = self.last_char.take().into_iter().chain(&mut self.chars);
```

Comment:
char_iter

Reply:
I'm not sure what you mean by this

Reply:
A sorry, I meant `iter` is not a great name.
`char_iter` would be nicer.

You can also put explicit types when it is useful like this.

`let mut char_it: impl Iterator<Item=char> = `